readthedocs/integrations/models.py

```diff
@@ -0,0 +1,145 @@
+from django.contrib.contenttypes.models import ContentType
+from django.utils.safestring import mark_safe
+from rest_framework import status
+from jsonfield import JSONField
+from pygments import highlight
+from pygments.lexers import JsonLexer
+from pygments.formatters import HtmlFormatter
+
+
+class HttpTransactionManager(models.Manager):
+
+    """HTTP transaction manager methods"""
+
+    @transaction.atomic
+    def from_transaction(self, req, resp, related_object, payload=None):
+        """Create object from Django request and response objects
+
+        If an explicit Request ``payload`` is not specified, the payload will be
+        determined directly from the Request object. This makes a good effort to
+        normalize the data, however we don't enforce that the payload is JSON
+
+        :param req: Request object to store
+        :type req: HttpRequest
+        :param resp: Response object to store
+        :type resp: HttpResponse
+        :param related_object: Object to use for generic relation
+        :param payload: Alternate payload object to store
+        :type payload: dict
+        """
+        request_payload = payload
+        if request_payload is None:
+            request_payload = getattr(req, 'data', None)
+            if req.content_type != 'application/json':
+                # Here, request_body can be a dict or a MergeDict. Probably best
+                # to normalize everything first
+                try:
+                    request_payload = dict(request_payload.items())
+                except AttributeError:
+                    pass
+        try:
+            request_body = json.dumps(request_payload, sort_keys=True)
+        except TypeError:
+            request_body = str(request_payload)
+        # This is the rawest form of request header we have, the WSGI
+        # headers. HTTP headers are prefixed with `HTTP_`, which we remove,
+        # and because the keys are all uppercase, we'll normalize them to
+        # title case-y hyphen separated values.
+        request_headers = dict(
+            (key[5:].title().replace('_', '-'), str(val))
+            for (key, val) in req.META.items()
+            if key.startswith('HTTP_')
+        )
+        request_headers['Content-Type'] = req.content_type
+
+        response_payload = resp.data if hasattr(resp, 'data') else resp.content
+        try:
+            response_body = json.dumps(response_payload, sort_keys=True)
+        except TypeError:
+            response_body = str(response_payload)
+        response_headers = dict(resp.items())
+
+        fields = {
+            'status_code': resp.status_code,
+            'request_headers': request_headers,
+            'request_body': request_body,
+            'response_body': response_body,
+            'response_headers': response_headers,
+        }
+        fields['related_object'] = related_object
+        self.delete_limit(related_object, limit=9)
```

Comment:
Only because we are cleaning before the object is added, I figured there might be another case where we want to call this clean up, and that default should be 10. I could alter this to clean after the object is created if it makes a difference.